core/trie_verify.go

```diff
@@ -0,0 +1,345 @@
+package core
+
+import (
+	"fmt"
+	"math/rand"
+	"time"
+
+	lru "github.com/hashicorp/golang-lru"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+const (
+	verifiedCacheSize = 256
+	maxForkHeight     = 11
+
+	// defaultPeerNumber is default number of verify peers
+	defaultPeerNumber = 3
+	// pruneHeightDiff indicates that if the height difference between current block and task's
+	// corresponding block is larger than it, the task should be pruned.
+	pruneHeightDiff = 15
+	pruneInterval   = 5 * time.Second
+	resendInterval  = 2 * time.Second
+	// tryAllPeersTime is the time that a block has not been verified and then try all the valid verify peers.
+	tryAllPeersTime = 15 * time.Second
+)
+
+type verifyManager struct {
+	bc             *BlockChain
+	tasks          map[common.Hash]*verifyTask
+	peers          verifyPeers
+	verifiedCache  *lru.Cache
+	allowUntrusted bool
+
+	newTaskCh chan *types.Header
+	verifyCh  chan common.Hash
+	messageCh chan verifyMessage
+}
+
+func NewVerifyManager(blockchain *BlockChain, peers verifyPeers, allowUntrusted bool) *verifyManager {
+	verifiedCache, _ := lru.New(verifiedCacheSize)
+	vm := &verifyManager{
+		bc:             blockchain,
+		tasks:          make(map[common.Hash]*verifyTask),
+		peers:          peers,
+		verifiedCache:  verifiedCache,
+		allowUntrusted: allowUntrusted,
+
+		newTaskCh: make(chan *types.Header),
+		verifyCh:  make(chan common.Hash),
```

Comment:
I will suggest `verifyCh` have some buffer, like 11, so than the main process will not been blocked.

Reply:
So do `newTaskCh`, please allocate some buffer.